File: code/java/algorithm_thinking_article-17.java
public class Algorithm_thinking_article_17 {
    /**
     * Backtracking algorithm practice: set partitioning
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 2) {
            throw new IllegalArgumentException("Expected: (int[] nums, int k)");
        }

        Solution solution = new Solution();
        return solution.canPartitionKSubsets((int[]) args[0], (int) args[1]);
    }

    private static class Solution {

        public boolean canPartitionKSubsets(int[] nums, int k) {
            if (k > nums.length) return false;
            int sum = 0;
            for (int v : nums) sum += v;
            if (sum % k != 0) return false;
            int used = 0;
            int target = sum / k;
            return backtrack(k, 0, nums, 0, used, target);
        }
        HashMap<Integer, Boolean> memo = new HashMap<>();
        boolean backtrack(int k, int bucket,
                        int[] nums, int start, int used, int target) {
            if (k == 0) {
                return true;
            }
            if (bucket == target) {
                boolean res = backtrack(k - 1, 0, nums, 0, used, target);
                memo.put(used, res);
                return res;
            }
            if (memo.containsKey(used)) {
                return memo.get(used);
            }
            for (int i = start; i < nums.length; i++) {
                if (((used >> i) & 1) == 1) {
                    continue;
                }
                if (nums[i] + bucket > target) {
                    continue;
                }
                used |= 1 << i;
                bucket += nums[i];
                if (backtrack(k, bucket, nums, i + 1, used, target)) {
                    return true;
                }
                used ^= 1 << i;
                bucket -= nums[i];
            }
            return false;
        }
    }
}

File: code/python/algorithm_thinking_article-17.py
"""Backtracking algorithm practice: set partitioning"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().canPartitionKSubsets(*args, **kwargs)


class _Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        if k > len(nums):
            return False
        _sum = sum(nums)
        if _sum % k != 0:
            return False
        used = 0
        target = _sum
        return self.backtrack(k, 0, nums, 0, used, target, {})
    def backtrack(self, k: int, bucket: int, nums: List[int], start: int, used: int, target: int, memo: Dict[int, bool]) -> bool:
        if k == 0:
            return True
        if bucket == target:
            if used in memo:
                return memo[used]
            res = self.backtrack(k - 1, 0, nums, 0, used, target, memo)
            memo[used] = res
            return res
        for i in range(start, len(nums)):
            if used >> i & 1 == 1:
                continue
            if nums[i] + bucket > target:
                continue
            used |= 1 << i
            bucket += nums[i]
            if self.backtrack(k, bucket, nums, i + 1, used, target, memo):
                return True
            used ^= 1 << i
            bucket -= nums[i]
        return False
